File: fusion/crates/fusion-remote/src/bootstrap.rs
//! Password → key authentication bootstrap.
//!
//! The rest of fusion-remote uses OpenSSH's ControlMaster pipeline with
//! `BatchMode=yes`, which **does not accept passwords**. This module makes
//! the first connection ergonomic for users who only know a password:
//!
//! 1. Pick or generate a managed ed25519 keypair on the scanner host
//!    (default `~/.config/scdr/fusion-remote/keys/<user>@<host>-<port>.ed25519`).
//! 2. Try key-based login (via `ssh2`). If it works, return that path.
//! 3. Otherwise use the password (via `ssh2`) to append the public key to
//!    the remote `~/.ssh/authorized_keys`, then verify key login.
//! 4. Hand the verified key path back; everything downstream uses keys
//!    only — the password is dropped from memory.

use std::fs;
use std::io::Read;
use std::net::{TcpStream, ToSocketAddrs};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::Duration;

use anyhow::{anyhow, bail, Context};
use ssh2::Session;

use crate::sh_quote;

const HANDSHAKE_TIMEOUT_MS: u32 = 10_000;
const CONNECT_TIMEOUT: Duration = Duration::from_secs(10);

/// Default managed-key path on the scanner host.
pub fn default_key_path(user: &str, host: &str, port: u16) -> anyhow::Result<PathBuf> {
    let base = dirs::config_dir().ok_or_else(|| anyhow!("cannot resolve user config dir"))?;
    let dir = base.join("scdr/fusion-remote/keys");
    fs::create_dir_all(&dir).with_context(|| format!("create {}", dir.display()))?;
    let name = format!("{}@{}-{port}.ed25519", sanitize(user), sanitize(host));
    Ok(dir.join(name))
}

/// Ensure key-based SSH login works against `user@host:port`, installing
/// `~/.ssh/authorized_keys` over a one-shot password session if needed.
///
/// Returns the verified private key path.
pub fn ensure_key_auth(
    target: &str,
    port: u16,
    identity: Option<&Path>,
    password: Option<&str>,
) -> anyhow::Result<PathBuf> {
    let (user, host) = parse_target(target)?;

    let key = match identity {
        Some(p) => p.to_path_buf(),
        None => default_key_path(&user, &host, port)?,
    };
    let pub_key = pub_path(&key);

    if !key.exists() {
        generate_keypair(&key, &user, &host)
            .with_context(|| format!("generate ed25519 keypair at {}", key.display()))?;
    } else if !pub_key.exists() {
        bail!(
            "private key {} exists but matching .pub file is missing",
            key.display()
        );
    }

    if key_auth_succeeds(&user, &host, port, &key)? {
        return Ok(key);
    }

    let pw = password.ok_or_else(|| {
        anyhow!(
            "key authentication failed for {user}@{host}:{port} and no password \
             provided; pass --ssh-password / --ssh-password-stdin / set \
             SCDR_SSH_PASSWORD env"
        )
    })?;

    let pub_line =
        fs::read_to_string(&pub_key).with_context(|| format!("read {}", pub_key.display()))?;
    install_public_key(&host, port, &user, pw, pub_line.trim())
        .context("install public key via password ssh")?;

    if !key_auth_succeeds(&user, &host, port, &key)? {
        bail!(
            "public key appended to authorized_keys but key login still \
             fails on {user}@{host}:{port}"
        );
    }
    Ok(key)
}

fn parse_target(t: &str) -> anyhow::Result<(String, String)> {
    let (u, h) = t
        .split_once('@')
        .ok_or_else(|| anyhow!("ssh target must be user@host, got {t:?}"))?;
    if u.is_empty() || h.is_empty() {
        bail!("ssh target has empty user or host: {t:?}");
    }
    Ok((u.to_string(), h.to_string()))
}

fn sanitize(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.') {
                c
            } else {
                '_'
            }
        })
        .collect()
}

fn generate_keypair(path: &Path, user: &str, host: &str) -> anyhow::Result<()> {
    // ssh-keygen will not overwrite, but path is guaranteed not to exist.
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).with_context(|| format!("create {}", parent.display()))?;
    }
    let out = Command::new("ssh-keygen")
        .args(["-t", "ed25519", "-N", "", "-q"])
        .arg("-f")
        .arg(path)
        .arg("-C")
        .arg(format!("fusion-remote@{user}@{host}"))
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .output()
        .context("spawn ssh-keygen (is openssh-client installed?)")?;
    if !out.status.success() {
        bail!(
            "ssh-keygen failed: {}",
            String::from_utf8_lossy(&out.stderr).trim()
        );
    }
    Ok(())
}

fn open_session(host: &str, port: u16, timeout: Duration) -> anyhow::Result<Session> {
    let addr = (host, port)
        .to_socket_addrs()
        .with_context(|| format!("resolve {host}:{port}"))?
        .next()
        .ok_or_else(|| anyhow!("no address for {host}:{port}"))?;
    let tcp = TcpStream::connect_timeout(&addr, timeout)
        .with_context(|| format!("tcp connect {addr}"))?;
    tcp.set_read_timeout(Some(timeout))?;
    tcp.set_write_timeout(Some(timeout))?;
    let mut sess = Session::new().context("ssh2 session")?;
    sess.set_tcp_stream(tcp);
    sess.set_timeout(HANDSHAKE_TIMEOUT_MS);
    sess.handshake().context("ssh2 handshake")?;
    Ok(sess)
}

fn key_auth_succeeds(user: &str, host: &str, port: u16, key: &Path) -> anyhow::Result<bool> {
    let sess = match open_session(host, port, Duration::from_secs(5)) {
        Ok(s) => s,
        Err(_) => return Ok(false),
    };
    let pub_key = pub_path(key);
    let pub_arg = if pub_key.exists() {
        Some(pub_key.as_path())
    } else {
        None
    };
    Ok(sess
        .userauth_pubkey_file(user, pub_arg, key, None)
        .map(|()| sess.authenticated())
        .unwrap_or(false))
}

fn install_public_key(
    host: &str,
    port: u16,
    user: &str,
    password: &str,
    pubkey_line: &str,
) -> anyhow::Result<()> {
    let sess = open_session(host, port, CONNECT_TIMEOUT)?;
    sess.userauth_password(user, password)
        .context("ssh password authentication")?;
    if !sess.authenticated() {
        bail!("ssh password auth ok but session not authenticated");
    }

    let q = sh_quote(pubkey_line);
    let cmd = format!(
        "mkdir -p ~/.ssh && chmod 700 ~/.ssh && \
         touch ~/.ssh/authorized_keys && chmod 600 ~/.ssh/authorized_keys && \
         (grep -qxF {q} ~/.ssh/authorized_keys || echo {q} >> ~/.ssh/authorized_keys)"
    );

    let mut channel = sess.channel_session().context("open exec channel")?;
    channel.exec(&cmd).context("exec authorized_keys install")?;

    let mut stdout = String::new();
    let _ = channel.read_to_string(&mut stdout);
    let mut stderr = String::new();
    let _ = channel.stderr().read_to_string(&mut stderr);
    channel.wait_close().context("wait_close exec channel")?;
    let exit = channel
        .exit_status()
        .context("read exit_status from exec channel")?;
    if exit != 0 {
        bail!(
            "authorized_keys install exit {exit}\nstdout: {}\nstderr: {}",
            stdout.trim(),
            stderr.trim()
        );
    }
    Ok(())
}

/// `ssh-keygen` writes the public key as `<priv>.pub` — appended, not
/// extension-replaced. `Path::with_extension` is wrong here.
fn pub_path(key: &Path) -> PathBuf {
    let mut p = key.as_os_str().to_owned();
    p.push(".pub");
    PathBuf::from(p)
}

/// Resolve the tool-managed private-key path for `target` (`user@host`) and
/// `port` — the key [`ensure_key_auth`] generates/uses when no explicit
/// identity is given. Useful for cleaning up local state after a revoke.
pub fn managed_key_path(target: &str, port: u16) -> anyhow::Result<PathBuf> {
    let (user, host) = parse_target(target)?;
    default_key_path(&user, &host, port)
}

/// Remove the managed public key this tool installed from the target's
/// `~/.ssh/authorized_keys`, and report whether a line was actually removed
/// (`false` = already absent).
///
/// The inverse of [`ensure_key_auth`]'s key install: it leaves the target's
/// SSH setup as it was before bootstrap. To minimize footprint, **only the
/// exact key line is removed** (whole-line fixed-string match) — every other
/// authorized key is untouched, the file keeps its mode/owner, and `~/.ssh`
/// and `authorized_keys` are left in place (we cannot know whether they
/// pre-existed). Authenticates with the managed key when it still works,
/// falling back to `password`.
pub fn revoke_key(
    target: &str,
    port: u16,
    identity: Option<&Path>,
    password: Option<&str>,
) -> anyhow::Result<bool> {
    let (user, host) = parse_target(target)?;
    let key = match identity {
        Some(p) => p.to_path_buf(),
        None => default_key_path(&user, &host, port)?,
    };
    let pub_key = pub_path(&key);
    let pub_line = fs::read_to_string(&pub_key)
        .with_context(|| format!("read managed public key {}", pub_key.display()))?
        .trim()
        .to_string();
    if pub_line.is_empty() {
        bail!("managed public key {} is empty", pub_key.display());
    }

    let sess = authenticated_session(&user, &host, port, &key, password)?;
    remove_authorized_key(&sess, &pub_line)
}

/// Open an ssh2 session authenticated by `key` if it still works, else by
/// `password`. A session that failed auth cannot be reused, so password auth
/// opens a fresh connection.
fn authenticated_session(
    user: &str,
    host: &str,
    port: u16,
    key: &Path,
    password: Option<&str>,
) -> anyhow::Result<Session> {
    if key.exists() {
        let sess = open_session(host, port, CONNECT_TIMEOUT)?;
        let pub_key = pub_path(key);
        let pub_arg = if pub_key.exists() {
            Some(pub_key.as_path())
        } else {
            None
        };
        if sess
            .userauth_pubkey_file(user, pub_arg, key, None)
            .map(|()| sess.authenticated())
            .unwrap_or(false)
        {
            return Ok(sess);
        }
    }

    let pw = password.ok_or_else(|| {
        anyhow!(
            "cannot authenticate to {user}@{host}:{port} to revoke: managed key \
             auth failed and no password provided (pass --ssh-password or set \
             SCDR_SSH_PASSWORD)"
        )
    })?;
    let sess = open_session(host, port, CONNECT_TIMEOUT)?;
    sess.userauth_password(user, pw)
        .context("ssh password authentication for revoke")?;
    if !sess.authenticated() {
        bail!("ssh password auth ok but session not authenticated");
    }
    Ok(sess)
}

/// Delete exactly `pub_line` from the remote `~/.ssh/authorized_keys`,
/// preserving the file's mode and owner and leaving every other entry intact.
/// Returns `false` when the file or line is absent (nothing changed).
fn remove_authorized_key(sess: &Session, pub_line: &str) -> anyhow::Result<bool> {
    let q = sh_quote(pub_line);
    // Whole-line, fixed-string match (`grep -xF`) so only our exact entry is
    // touched. Rewrite through a temp file in the same directory, then copy the
    // bytes back into the original file so it keeps its inode, mode, and owner.
    let cmd = format!(
        "f=\"$HOME/.ssh/authorized_keys\"; \
         [ -f \"$f\" ] || {{ echo __absent; exit 0; }}; \
         if grep -qxF {q} \"$f\"; then \
           tmp=$(mktemp \"$f.XXXXXX\") || exit 1; \
           grep -vxF {q} \"$f\" > \"$tmp\" || true; \
           cat \"$tmp\" > \"$f\"; rm -f \"$tmp\"; \
           echo __removed; \
         else echo __absent; fi"
    );
    let (stdout, stderr, code) = exec_capture(sess, &cmd)?;
    if code != 0 {
        bail!("revoke command failed (exit {code}): {}", stderr.trim());
    }
    Ok(stdout.contains("__removed"))
}

/// Run `cmd` over an authenticated ssh2 session; capture stdout, stderr, exit.
fn exec_capture(sess: &Session, cmd: &str) -> anyhow::Result<(String, String, i32)> {
    let mut channel = sess.channel_session().context("open exec channel")?;
    channel.exec(cmd).context("exec remote command")?;
    let mut stdout = String::new();
    channel.read_to_string(&mut stdout).ok();
    let mut stderr = String::new();
    channel.stderr().read_to_string(&mut stderr).ok();
    channel.wait_close().context("wait_close exec channel")?;
    let code = channel.exit_status().context("read exit_status")?;
    Ok((stdout, stderr, code))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_target_basic() {
        assert_eq!(
            parse_target("root@10.0.0.1").unwrap(),
            ("root".into(), "10.0.0.1".into())
        );
    }

    #[test]
    fn parse_target_rejects_bad() {
        assert!(parse_target("no-at").is_err());
        assert!(parse_target("@host").is_err());
        assert!(parse_target("user@").is_err());
    }

    #[test]
    fn sanitize_replaces_path_chars() {
        assert_eq!(sanitize("user/with:dots."), "user_with_dots.");
        assert_eq!(sanitize("OK_42"), "OK_42");
    }

    #[test]
    fn pub_path_appends_suffix_not_replace() {
        let key = Path::new("/tmp/k.ed25519");
        assert_eq!(pub_path(key), Path::new("/tmp/k.ed25519.pub"));
        let key2 = Path::new("/tmp/k");
        assert_eq!(pub_path(key2), Path::new("/tmp/k.pub"));
    }

    #[test]
    fn default_key_path_uses_sanitized_names() {
        let p = default_key_path("u/x", "h:1", 22).unwrap();
        let name = p.file_name().unwrap().to_string_lossy().into_owned();
        assert_eq!(name, "u_x@h_1-22.ed25519");
    }
}

File: form/src/kcatta_form/deploy/bootstrap.py
"""Form-owned password -> key SSH authentication bootstrap.

The scan pipeline authenticates with keys (paramiko, key only). This module
makes the first connection ergonomic for operators who only know a password:

1. Pick / generate a managed ed25519 keypair on the scanner host. Its filename
   combines a readable target label with a digest of the original
   ``(user, host, port)`` identity, so lossy filename sanitization cannot make
   two targets share one key.
2. Try key login. If it works, return that path.
3. Otherwise use the password once to append the public key to the target's
   ``~/.ssh/authorized_keys``, verify key login, then drop the password.

``revoke_key`` is the inverse: it removes exactly the line this tool added.
"""

from __future__ import annotations

import contextlib
import hashlib
import json
import os
import re
import shutil
import subprocess
from pathlib import Path

import paramiko

from ._util import sh_quote, split_user_host
from .ssh import create_ssh_client

CONNECT_TIMEOUT = 10.0


def _sanitize(value: str) -> str:
    # Keep the readable portion single-byte and filesystem-portable; the digest
    # below, not this lossy label, carries the complete Unicode identity.
    return "".join(c if (c.isascii() and (c.isalnum() or c in "-_.")) else "_" for c in value)


def _managed_path_stem(user: str, host: str, port: int) -> str:
    """Collision-resistant filename stem for a managed target credential.

    ``_sanitize`` is intentionally lossy (for example, ``DOMAIN\\user`` and
    ``DOMAIN_user`` sanitize identically). Keep its readable label, but bind the
    filename to the unsanitized logical identity with a full SHA-256 digest.
    Canonical JSON keeps the three typed input fields unambiguous.
    """
    identity = json.dumps([user, host, port], ensure_ascii=False, separators=(",", ":")).encode(
        "utf-8"
    )
    digest = hashlib.sha256(identity).hexdigest()
    # NAME_MAX is commonly 255 bytes. Bound the readable prefix so maximal DNS
    # names still leave room for the digest and `.ed25519` / `.crt` suffixes.
    readable_user = _sanitize(user)[:40]
    readable_host = _sanitize(host)[:80]
    return f"{readable_user}@{readable_host}-{port}-{digest}"


def _config_dir() -> Path:
    base = os.environ.get("XDG_CONFIG_HOME")
    return Path(base) if base else Path.home() / ".config"


def _ensure_key_dir() -> Path:
    """The managed-keys dir, created/tightened to 0o700 (private to the Form user).

    ``mkdir(mode=…)`` does NOT tighten a pre-existing dir (and ``exist_ok=True``
    skips the mode entirely), so chmod the leaf explicitly — otherwise a dir left
    at the default 0o755 by an earlier version would stay group/world-traversable,
    letting other local users enumerate which user@host:port targets exist.
    """
    directory = _config_dir() / "scdr" / "agent-remote" / "keys"
    directory.mkdir(parents=True, exist_ok=True, mode=0o700)
    with contextlib.suppress(OSError):
        directory.chmod(0o700)
    return directory


def default_key_path(user: str, host: str, port: int) -> Path:
    """Managed-key path on the scanner host (created if missing)."""
    directory = _ensure_key_dir()
    return directory / f"{_managed_path_stem(user, host, port)}.ed25519"


def managed_key_path(target: str, port: int) -> Path:
    """Resolve the tool-managed private-key path for ``user@host`` + ``port``."""
    user, host = split_user_host(target)
    return default_key_path(user, host, port)


def _pub_path(key: Path) -> Path:
    # ssh-keygen appends `.pub` (it does not replace the extension).
    return key.with_name(key.name + ".pub")


def _generate_keypair(path: Path, user: str, host: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    result = subprocess.run(
        [
            "ssh-keygen",
            "-t",
            "ed25519",
            "-N",
            "",
            "-q",
            "-f",
            str(path),
            "-C",
            f"agent-remote@{user}@{host}",
        ],
        stdin=subprocess.DEVNULL,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"ssh-keygen failed (is openssh-client installed?): {result.stderr.strip()}"
        )
    # ssh-keygen already 0o600s the private key; tighten the .pub too so a
    # multi-tenant Form host can't leak public keys/fingerprints by readdir.
    with contextlib.suppress(OSError):
        _pub_path(path).chmod(0o600)


def _key_auth_succeeds(user: str, host: str, port: int, key: Path) -> bool:
    client = create_ssh_client()
    try:
        client.connect(
            hostname=host,
            port=port,
            username=user,
            key_filename=str(key),
            look_for_keys=False,
            allow_agent=False,
            timeout=5.0,
        )
        return True
    except Exception:
        return False
    finally:
        client.close()


def _password_session(
    user: str, host: str, port: int, password: str | None = None, key: Path | None = None
) -> paramiko.SSHClient:
    """Open an SSH session — by password, or by managed key when ``key`` is given.

    Password bootstrap, key probes, scans, rotate, and revoke all use the same
    persistent known_hosts configuration from :func:`create_ssh_client`.
    """
    client = create_ssh_client()
    if key is not None:
        client.connect(
            hostname=host,
            port=port,
            username=user,
            key_filename=str(key),
            look_for_keys=False,
            allow_agent=False,
            timeout=CONNECT_TIMEOUT,
        )
    else:
        client.connect(
            hostname=host,
            port=port,
            username=user,
            password=password,
            look_for_keys=False,
            allow_agent=False,
            timeout=CONNECT_TIMEOUT,
        )
    return client


def _key_session(user: str, host: str, port: int, key: Path) -> paramiko.SSHClient:
    """Open a key-authenticated session (delegates to the shared session opener)."""
    return _password_session(user, host, port, key=key)


def _authed_session(
    user: str, host: str, port: int, key: Path, password: str | None
) -> paramiko.SSHClient:
    """Authenticate preferring the managed key, falling back to a one-time password.

    Used by rotate/revoke, which need an authenticated session to edit the
    target's ``authorized_keys``: when the current managed key still works no
    password is needed; otherwise the supplied ``password`` is used once. Raises
    if neither can authenticate.
    """
    if key.exists() and _key_auth_succeeds(user, host, port, key):
        return _key_session(user, host, port, key)
    if password:
        return _password_session(user, host, port, password)
    raise RuntimeError(
        f"cannot authenticate to {user}@{host}:{port}: managed key auth failed "
        "and no password was provided"
    )


def _exec(client: paramiko.SSHClient, command: str) -> tuple[str, str, int]:
    _stdin, stdout, stderr = client.exec_command(command)
    out = stdout.read().decode("utf-8", "replace")
    err = stderr.read().decode("utf-8", "replace")
    status = stdout.channel.recv_exit_status()
    return out, err, status


def _install_pub_cmd(pub_line: str) -> str:
    """Idempotently append a public key line to the target's authorized_keys."""
    quoted = sh_quote(pub_line)
    return (
        "mkdir -p ~/.ssh && chmod 700 ~/.ssh && "
        "touch ~/.ssh/authorized_keys && chmod 600 ~/.ssh/authorized_keys && "
        f"(grep -qxF {quoted} ~/.ssh/authorized_keys || echo {quoted} >> ~/.ssh/authorized_keys)"
    )


def _remove_pub_cmd(pub_line: str) -> str:
    """Remove the exact public-key identity from ``authorized_keys``.

    Match the algorithm + base64 blob, not the mutable comment or key options.
    Rewrites through a temp file in the same dir and verifies the key blob is
    absent after the atomic rename. Prints ``__removed`` / ``__absent``.
    """
    fields = pub_line.split()
    if len(fields) < 2:
        raise RuntimeError("managed SSH public key is malformed")
    key_type, key_blob = fields[0], fields[1]
    if not re.fullmatch(r"[A-Za-z0-9@._+-]+", key_type) or not re.fullmatch(
        r"[A-Za-z0-9+/=]+", key_blob
    ):
        raise RuntimeError("managed SSH public key contains invalid algorithm/blob characters")
    quoted_type = sh_quote(key_type)
    quoted_blob = sh_quote(key_blob)
    matcher = (
        f"awk -v kt={quoted_type} -v kb={quoted_blob} "
        "'{ for (i=1; i<NF; i++) if ($i == kt && $(i+1) == kb) found=1 } "
        "END { exit(found ? 0 : 1) }'"
    )
    filterer = (
        f"awk -v kt={quoted_type} -v kb={quoted_blob} "
        "'{ drop=0; for (i=1; i<NF; i++) if ($i == kt && $(i+1) == kb) drop=1; "
        "if (!drop) print }'"
    )
    return (
        'f="$HOME/.ssh/authorized_keys"; '
        '[ -f "$f" ] || { echo __absent; exit 0; }; '
        '[ -r "$f" ] || exit 1; '
        f'{matcher} "$f"; rc=$?; '
        'if [ "$rc" -eq 0 ]; then '
        '  tmp=$(mktemp "$f.XXXXXX") || exit 1; '
        '  chmod 600 "$tmp"; '
        f'  {filterer} "$f" > "$tmp" || {{ rm -f "$tmp"; exit 1; }}; '
        # Atomic replace: rename within the same dir so an interrupted revoke can
        # never leave authorized_keys truncated/empty (SSH lockout). mktemp is
        # 0600, matching the file mode.
        '  mv -f "$tmp" "$f" || exit 1; '
        f'  {matcher} "$f"; post_rc=$?; '
        '  if [ "$post_rc" -eq 0 ]; then echo __still_present >&2; exit 1; fi; '
        '  [ "$post_rc" -eq 1 ] || exit "$post_rc"; echo __removed; '
        'elif [ "$rc" -eq 1 ]; then echo __absent; '
        'else exit "$rc"; fi'
    )


def ensure_key_auth(
    target: str,
    port: int,
    identity: Path | None = None,
    password: str | None = None,
) -> Path:
    """Ensure key-based login works against ``user@host:port``; return the key path.

    Installs the managed public key over a one-shot password session when key
    auth is not yet set up.
    """
    user, host = split_user_host(target)
    key = identity if identity is not None else default_key_path(user, host, port)
    pub_key = _pub_path(key)

    if not key.exists():
        _generate_keypair(key, user, host)
    elif not pub_key.exists():
        raise RuntimeError(f"private key {key} exists but {pub_key} is missing")

    if _key_auth_succeeds(user, host, port, key):
        return key

    if not password:
        raise RuntimeError(
            f"key authentication failed for {user}@{host}:{port} and no password "
            "provided (pass --ssh-password / --ssh-password-stdin / set SCDR_SSH_PASSWORD)"
        )

    pub_line = pub_key.read_text(encoding="utf-8").strip()
    install_cmd = _install_pub_cmd(pub_line)
    client = _password_session(user, host, port, password)
    try:
        _out, err, code = _exec(client, install_cmd)
    finally:
        client.close()
    if code != 0:
        raise RuntimeError(f"authorized_keys install failed (exit {code}): {err.strip()}")

    if not _key_auth_succeeds(user, host, port, key):
        raise RuntimeError(f"public key appended but key login still fails on {user}@{host}:{port}")
    return key


def revoke_key(
    target: str,
    port: int,
    identity: Path | None = None,
    password: str | None = None,
) -> bool:
    """Remove exactly the managed public key line from the target's
    ``authorized_keys``. Returns ``True`` if a line was removed, ``False`` if
    it was already absent."""
    user, host = split_user_host(target)
    key = identity if identity is not None else default_key_path(user, host, port)
    pub = _pub_path(key)
    if not pub.exists():
        # `cred.exists` is gated on the PRIVATE key; a present-private / missing-pub
        # state (manual deletion, interrupted keygen) would otherwise crash here
        # with an opaque FileNotFoundError → 502. Fail with a clear message instead.
        raise RuntimeError(
            f"managed public key {pub} is missing; cannot determine the line to revoke"
        )
    pub_line = pub.read_text(encoding="utf-8").strip()
    if not pub_line:
        raise RuntimeError(f"managed public key {pub} is empty")

    client = _authed_session(user, host, port, key, password)
    try:
        out, err, code = _exec(client, _remove_pub_cmd(pub_line))
    finally:
        client.close()
    if code != 0:
        raise RuntimeError(f"revoke command failed (exit {code}): {err.strip()}")
    return "__removed" in out


def can_authenticate(target: str, port: int, identity: Path | None = None) -> bool:
    """True if the managed (or ``identity``) key can currently log into ``user@host:port``.

    Used by the credential-management "test connectivity" action: resolves the
    same key path the scan pipeline would use and probes a key-only login.
    """
    user, host = split_user_host(target)
    key = identity if identity is not None else default_key_path(user, host, port)
    if not key.exists():
        return False
    return _key_auth_succeeds(user, host, port, key)


def key_fingerprint(key: Path) -> str | None:
    """SHA256 fingerprint of the managed public key (e.g. ``SHA256:…``), or None.

    Reads ``<key>.pub`` via ``ssh-keygen -lf``. Returns None when the public key is
    absent or ssh-keygen is unavailable — a missing fingerprint is "unknown", not
    an error (the key may still be usable; the UI just shows it as unverified).
    """
    pub = _pub_path(key)
    if not pub.exists():
        return None
    try:
        # Pin -E sha256 so the format is deterministic regardless of the host's
        # FingerprintHash default (schema/UI promise a SHA256 fingerprint).
        result = subprocess.run(
            ["ssh-keygen", "-lf", str(pub), "-E", "sha256"],
            stdin=subprocess.DEVNULL,
            capture_output=True,
            text=True,
        )
    except OSError:
        return None
    if result.returncode != 0:
        return None
    # Output: "<bits> SHA256:<hash> <comment> (<TYPE>)" — pull the hash token.
    for token in result.stdout.split():
        if token.startswith("SHA256:"):
            return token
    return None


def rotate_key(target: str, port: int, password: str | None = None) -> Path:
    """Rotate the tool-managed SSH key for ``user@host:port``; return the key path.

    Generates a fresh keypair, installs the new public key — authenticating with
    the *current* managed key when it still works (no password needed), else a
    one-time ``password`` — verifies the new key logs in, removes the old public
    key line from the target's ``authorized_keys``, then atomically swaps the new
    keypair into the managed path. The old key is left untouched until the new one
    is proven working, so a failed rotation never locks Form out.
    """
    user, host = split_user_host(target)
    key = default_key_path(user, host, port)
    pub = _pub_path(key)
    if not key.is_file() or not pub.is_file():
        raise RuntimeError(
            f"managed SSH keypair for {user}@{host}:{port} is incomplete; "
            "bootstrap or recover both private and public key files before rotation"
        )
    old_pub_line = pub.read_text(encoding="utf-8").strip()
    if not old_pub_line:
        raise RuntimeError(f"managed SSH public key {pub} is empty; refusing unsafe rotation")

    # Stage the new keypair beside the managed one (same dir → atomic os.replace).
    new_key = key.with_name(key.name + ".new")
    new_pub = _pub_path(new_key)
    old_key_backup = key.with_name(key.name + ".rotate-old")
    old_pub_backup = pub.with_name(pub.name + ".rotate-old")
    if old_key_backup.exists() or old_pub_backup.exists():
        raise RuntimeError(
            f"incomplete prior SSH key rotation backup exists for {user}@{host}:{port}; "
            "recover or remove it before retrying"
        )
    new_key.unlink(missing_ok=True)
    new_pub.unlink(missing_ok=True)
    _generate_keypair(new_key, user, host)

    client: paramiko.SSHClient | None = None
    promoted = False
    try:
        new_pub_line = new_pub.read_text(encoding="utf-8").strip()
        # Authenticate with the OLD key (still in place) when possible, else password.
        client = _authed_session(user, host, port, key, password)
        _out, err, code = _exec(client, _install_pub_cmd(new_pub_line))
        if code != 0:
            raise RuntimeError(f"new authorized_keys install failed (exit {code}): {err.strip()}")

        if not _key_auth_succeeds(user, host, port, new_key):
            # The old credential is still active. Remove the staged remote key so
            # a failed proof does not leave an unmanaged authorized_keys entry.
            with contextlib.suppress(Exception):
                _exec(client, _remove_pub_cmd(new_pub_line))
            raise RuntimeError(
                f"new public key appended but key login still fails on {user}@{host}:{port}"
            )

        # Keep a rollback copy until both the local promotion and remote revoke
        # commit. The new key is proven before either old copy is touched.
        if key.exists():
            shutil.copy2(key, old_key_backup)
        if pub.exists():
            shutil.copy2(pub, old_pub_backup)
        try:
            os.replace(new_key, key)
            os.replace(new_pub, pub)
            promoted = True
        except Exception:
            if old_key_backup.exists():
                os.replace(old_key_backup, key)
            if old_pub_backup.exists():
                os.replace(old_pub_backup, pub)
            with contextlib.suppress(Exception):
                _exec(client, _remove_pub_cmd(new_pub_line))
            raise

        # Only revoke the old remote key after the new managed keypair is both
        # verified and locally active. A cleanup response can be ambiguous (the
        # remote rename may have committed just before the connection dropped),
        # so never roll back or remove the proven new key after this point.
        if old_pub_line and old_pub_line != new_pub_line:
            new_client: paramiko.SSHClient | None = None
            try:
                new_client = _key_session(user, host, port, key)
                _out, err, code = _exec(new_client, _remove_pub_cmd(old_pub_line))
                if code != 0:
                    raise RuntimeError(
                        f"old authorized_keys revoke failed (exit {code}): {err.strip()}"
                    )
            except Exception as cleanup_error:
                raise RuntimeError(
                    "new SSH key is active, but old authorized_keys cleanup failed; "
                    f"recovery copies remain at {old_key_backup} and {old_pub_backup}"
                ) from cleanup_error
            finally:
                if new_client is not None:
                    new_client.close()

        old_key_backup.unlink(missing_ok=True)
        old_pub_backup.unlink(missing_ok=True)
        return key
    except Exception:
        if not promoted:
            new_key.unlink(missing_ok=True)
            new_pub.unlink(missing_ok=True)
        raise
    finally:
        if client is not None:
            client.close()
